homeassistant/components/improv_ble/config_flow.py

```diff
@@ -125,22 +115,66 @@ async def async_step_user(
             errors=errors,
         )
 
-    async def async_step_bluetooth(
-        self, discovery_info: BluetoothServiceInfoBleak
-    ) -> FlowResult:
-        """Handle the Bluetooth discovery step."""
-        await self.async_set_unique_id(discovery_info.address)
-        self._abort_if_unique_id_configured()
-        service_data = discovery_info.service_data
+    def _abort_if_provisioned(self) -> None:
+        """Check improv state and abort flow if needed."""
+        # mypy is not aware that we can't get here without having these set already
+        assert self._discovery_info is not None
+
+        service_data = self._discovery_info.service_data
         improv_service_data = ImprovServiceData.from_bytes(
             service_data[SERVICE_DATA_UUID]
         )
         if improv_service_data.state in (State.PROVISIONING, State.PROVISIONED):
             _LOGGER.debug(
-                "Device is already provisioned: %s", improv_service_data.state
+                "Aborting improv flow, device is already provisioned: %s",
+                improv_service_data.state,
             )
-            return self.async_abort(reason="already_provisioned")
+            raise AbortFlow("already_provisioned")
+
+    @callback
+    def _async_update_ble(
+        self,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        _LOGGER.debug(
+            "Got updated BLE data: %s",
+            service_info.service_data[SERVICE_DATA_UUID].hex(),
+        )
+
+        self._discovery_info = service_info
+        try:
+            self._abort_if_provisioned()
+        except AbortFlow:
+            self.hass.config_entries.flow.async_abort(self.flow_id)
```

Comment:
If we expect this to abort frequently, I'd probably rewrite this break it out to return true/false if its provisioned instead of controlling the happy path with an exception... but I think its probably rare that it happens at this stage?

Reply:
yeah, it must be very rare indeed, the most common case will be aborting when starting the `bluetooth` step